### ISSUE TITLE
Linker error compiling MediaPipe face detection on Windows

### ISSUE THREAD

**Issue by jlouw2** [google/mediapipe#3851]
Environment:
OS: Windows 11
Compiler: Visual Studio 2022
Programming language: C++, 64-bit build
Bazel version 5.3.2
OpenCV version 3.4.16
Python version 3.10

Build command:
bazel build -c opt --define MEDIAPIPE_DISABLE_GPU=1 --action_env PYTHON_BIN_PATH="C://Program Files//Python310//python.exe" mediapipe/examples/desktop/face_detection:face_detection_cpu 

Problem:
face_detection_options_registry.lo.lib(face_detection_options_lib.obj) : error LNK2019: unresolved external symbol "private: static class RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry::registration_token<class mediapipe::FaceDetectionOptions>" (??$registration_token@VFaceDetectionOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@0VRegistrationToken@3@B) referenced in function "??__E?$registration_token@VFaceDetectionOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ" (??__E?$registration_token@VFaceDetectionOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ)

inference_calculator_options_registry.lo.lib(inference_calculator_options_lib.obj) : error LNK2019: unresolved external symbol "private: static class RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry::registration_token<class mediapipe::InferenceCalculatorOptions>" (??$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@0VRegistrationToken@3@B) referenced in function "??__E?$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ" (??__E?$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ)

bazel-out\x64_windows-opt\bin\mediapipe\examples\desktop\face_detection\face_detection_cpu.exe : fatal error LNK1120: 2 unresolved externals

**Comment by limsb95** [comment]
I am encountering same error message building face mesh solution on windows.
Any solution?

**Comment by kuaashish** [comment]
Hi @jlouw2,
MediaPipe team has not tested any solution which is available on Windows11 and running MediaPipe on Windows is still experimental as mentioned [here](https://google.github.io/mediapipe/getting_started/install.html#installing-on-windows). If you succeed in building the MediaPipe in Windows11 it will be instructions to other users. Thank you!

**Comment by jlouw2** [comment]
Hi @kuaashish,
I will test on Windows 10, but I doubt that Windows 11 is the problem.  The fact that everything compiles and that there are only 2 linker errors points to Bazel missing a library. I tried adding --subcommands --verbose_failures , but that did not help me to find the problem. hello_world compiled fine.  On Ubuntu (22.04 and 22.10) I got multiple compilation errors, long before getting to linking. Which Linux distro and versions of software have been validated to work?

**Comment by kuaashish** [comment]
Hi @jlouw2,
Mediapipe has been tested on the ubuntu version mentioned in the guide [here](https://google.github.io/mediapipe/getting_started/install.html#installing-on-debian-and-ubuntu). Could you follow the documentation and let us know the issue still persists on the mentioned version.

**Comment by jinfagang** [comment]
I get same link error on windows11.

I don't understand here, why googler not able to resolve windows 11 build error???

**Comment by jinfagang** [comment]
2023 , windows11 support?

**Comment by diegomacario** [comment]
I get the same linker error on Windows 10.

**Comment by diegomacario** [comment]
I worked around this problem by commenting these two lines in `mediapipe\framework\tool\options_registry.h`:

```
template <class MessageT>
static const RegistrationToken registration_token;
```

And by commenting this in `mediapipe\framework\tool\options_lib_template.cc`:

```
namespace mediapipe {
// The protobuf descriptor for an options message type.
template <>
const RegistrationToken tool::OptionsRegistry::registration_token<
    MP_OPTION_TYPE_NS::MP_OPTION_TYPE_NAME> =
    tool::OptionsRegistry::Register(ReadFileDescriptorSet(
        std::string(kDescriptorContents, sizeof(kDescriptorContents) - 1)));
}  // namespace mediapipe
```

Once I did that, the MediaPipe Hands example built successfully and worked perfectly.

This is a hacky workaround, but hopefully it helps.

**Comment by jinfagang** [comment]
@diegomacario Hello, I found the hands can run normally, but holistic can not run in this way. Find a proper way to support holistic as well?

**Comment by smbape** [comment]
I think it is a problem specific to visual studio 2022 >= 17.4, not a windows problem.
If you have an older version of visual studio, you can try it.
I opened a issue to the developper community, but it didn't get much attention and was closed.
[Link to the issue](https://developercommunity.visualstudio.com/t/Linking-failed-after-1736/10227061?)
It there is a way to make the issue being fixed, I think mediapipe will compile again

**Comment by rajkundu** [comment]
> I think it is a problem specific to visual studio 2022 >= 17.4, not a windows problem.

I can verify this; I have one machine with Windows 11 and VS 2022 17.**3**.5 - MediaPipe compiles perfectly. I have another machine with Windows 11 and VS 2022 17.**4**.5 - I get this error, and MediaPipe won't create graphs. So no, it is not a Windows 11 issue. If anyone has any ideas, please share!

**Comment by rajkundu** [comment]
@smbape Thank you for your great detective work! I am sorry that Microsoft wrongly closed your issue.

I have filed another issue with an even more simplified minimal reproducible example [here](https://developercommunity.visualstudio.com/t/Static-Data-Member-Templates:-DLL-Linkin/10288174). Please feel free to check it out and let me know if I have missed anything. @smbape @jinfagang @jlouw2 If you can, please consider upvoting the issue on Microsoft's forum to ensure that we can build MediaPipe with future versions of VS!

**Comment by rajkundu** [comment]
I can confirm that the underlying bug in VS 2022 and has been fixed in version 17.6 Preview 1. So, the permanent solution is to avoid using VS 2022 17.4.x & 17.5.x and either use VS 2022 <=17.3 or >=17.6 upon release.
